### PR TITLE
Use the player's overridden hand as the default for digging instead of always the default ("") hand.

### PR DESCRIPTION
**Pull request by shacknetisp**
Currently, even if the player's hand item is faster at digging than the empty item (default hand, `""`), the empty item will be used to determine digging parameters for the empty slot and non-tool items unless it cannot dig a node, in which case the actual hand is used.

This doesn't come up much in minetest game or similar games, where there is only one default hand, the `""` item, as wielded items will simply inherit that item's capabilities if they don't have their own and it is identical to the empty slot. However, this can be a problem if the hand is set to a more powerful tool, as the `""` item will still be used to determine dig parameters if it is possible to dig with it, even if the parameters are slower than the new hand.

This PR fixes that by using the player's actual hand as the default for digging parameters.

This PR also adds a `sethand` test command to devtest, so that hands can easily be tested.

## How to test

With this PR, load up a devtest game.
Dig some dirt with your hand, observe the speed.
Use `/sethand basetools:shovel_stone` to set your new hand.
Dig some dirt with your new shovel-hand or with a non-tool item, observe the speed increase as if you were actually wielding a shovel.

To see the bug, reset away from this PR (your hand will remain set, this PR only adds a helper command to set it) and try digging dirt with your new shovel-hand or a non-tool item. The speed will be as slow as if you had not set your hand in the first place.

### CODE REVIEW

File: src/network/serverpackethandler.cpp [hunk 1]
```diff
@@ -1187,11 +1187,17 @@ void Server::handleCommand_Interact(NetworkPacket *pkt)
...
-			if (!params.diggable) {
-				params = getDigParams(m_nodedef->get(n).groups,
+
+			// Get the digging parameters for the hand.
+			DigParams hand_params = getDigParams(m_nodedef->get(n).groups,
 					&hand_item.getToolCapabilities(m_itemdef));
+
+			// If the tool can't dig or takes more time than the hand,
+			// use the hand instead.
+			if(!params.diggable || (hand_params.diggable && params.time >= hand_params.time)) {
```
Comment (SmallJoker) [inline src/network/serverpackethandler.cpp:1197]: Please adjust the cheat detection code accordingly, for cases where the digging times makes a big difference:
https://github.com/minetest/minetest/blob/dc165fe942bcc48d7dea0a7b722886937d9c6914/src/network/serverpackethandler.cpp#L1205-L1211

Comment (shacknetisp) [reply src/network/serverpackethandler.cpp:1197]: I don't know where any changes would be necessary.
The new hand digging times only occur when the legitimate hand can dig faster. If the client is older (or tries to trick the server for some reason) and still uses the tool digging times, it just seems like the digging time is *longer* than expected, which doesn't seem to cause problems from by observations and tests (though I certainly don't understand the anticheat system entirely).

A newer client on an older server could run into trouble, but that would require backporting or another solution which may need to be addressed.

Comment (SmallJoker) [reply src/network/serverpackethandler.cpp:1197]: Ah yes right.  Sorry, I don't know what I thought there. The code and what you wrote totally makes sense.

**Comment by 0-afflatus** [comment]
Not sure if I fully understand the point of this. I don't like it from a gameplay perspective. Players should have to figure out the correct tool to use, this could potentially mask issues. In my game I have a deliberately duff tool as a booby prize, so I'd want to be able to disable this feature. I don't really think this should be in the engine

**Comment by shacknetisp** [comment]
I have redone the PR so that now tools will not be overridden, only non-tool (no tool capability) items and the empty slot will be.

**Comment by 0-afflatus** [comment]
> I have redone the PR so that now tools will not be overridden, only non-tool (no tool capability) items and the empty slot will be.
Thank you, that shouldn't break my runcible nerf spork now.

**Comment by Extex101** [comment]
Well in my mod I depend on the fact that there's a difference between hand and tool speeds. I use a rolling pin to roll out dough, it takes long when you use the rolling pin, but quicker to pick it up by hand.

**Comment by shacknetisp** [comment]
This PR should not affect your usecase; if the player is wielding a tool and the tool can dig a node, even if the time is longer than the hand, the tool will still be used to dig (as it was previously.)

This PR just fixes some cases where the player's overridden hand was being ignored in favor of the default hand.

**Pull request closed without merging**